kroto-plus-coroutines/src/main/kotlin/com/github/marcoferrer/krotoplus/coroutines/call/FlowControl.kt

```diff
@@ -42,40 +43,48 @@ internal fun <T> CallStreamObserver<*>.applyInboundFlowControl(
     }
 }
 
+private typealias MessageHandler = suspend ActorScope<*>.() -> Unit
+
 internal fun <T> CoroutineScope.applyOutboundFlowControl(
     streamObserver: CallStreamObserver<T>,
     targetChannel: Channel<T>
 ){
-    val isOutboundJobRunning = AtomicBoolean()
+
+    val isCompleted = AtomicBoolean()
     val channelIterator = targetChannel.iterator()
-    streamObserver.setOnReadyHandler {
-        if(targetChannel.isClosedForReceive){
-            streamObserver.completeSafely()
-        }else if(
+    val messageHandlerBlock: MessageHandler = handler@ {
+        while(
             streamObserver.isReady &&
-            !targetChannel.isClosedForReceive &&
-            isOutboundJobRunning.compareAndSet(false, true)
+            channelIterator.hasNext()
         ){
-            launch(Dispatchers.Unconfined + CoroutineExceptionHandler { _, e ->
-                streamObserver.completeSafely(e)
-                targetChannel.close(e)
-            }) {
-                try{
-                    while(
-                        streamObserver.isReady &&
-                        !targetChannel.isClosedForReceive &&
-                        channelIterator.hasNext()
-                    ){
-                        val value = channelIterator.next()
-                        streamObserver.onNext(value)
-                    }
-                    if(targetChannel.isClosedForReceive){
-                        streamObserver.onCompleted()
-                    }
-                } finally {
-                    isOutboundJobRunning.set(false)
-                }
-            }
+            streamObserver.onNext(channelIterator.next())
+        }
+        if(targetChannel.isClosedForReceive && isCompleted.compareAndSet(false,true)){
+            streamObserver.onCompleted()
+            channel.close()
+        }
+    }
+
+    val messageHandlerActor = actor<MessageHandler>(
+        capacity = Channel.UNLIMITED,
```

Comment:
A limited capacity didn't work? I am wondering if it's possible that the channel becomes a memory leak if jobs are added faster then the worker consumes them.

Reply:
Thats a good catch. I must've changed it while debugging. I'll test it with the value reverted to `CONFLATED`.

This implementation is based off the native grpc util `StreamObservers.copyWithFlowControl()` which ensured each invocation of the onReadyHandler always ran. [source](https://github.com/grpc/grpc-java/blob/9111602d7c2a88b7f9db6772d7d6d8399e2ff96b/stub/src/main/java/io/grpc/stub/StreamObservers.java#L41)

Reply:
I am using `CONFLATED` and have no issues.